custom_components/kia_uvo/switch.py

```diff
@@ -0,0 +1,131 @@
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.device_registry import DeviceEntry, async_get as dr_get
+from homeassistant.helpers.entity import DeviceInfo
+from homeassistant.helpers.update_coordinator import CoordinatorEntity
+
+from .const import DOMAIN
+
+
+# ---- helpers ----
+def _vin_from_device(device: DeviceEntry) -> str | None:
+    for dom, ident in device.identifiers:
+        if dom == DOMAIN:
+            return ident
+    return None
+
+
+async def _devices_for_entry(
+    hass: HomeAssistant, entry: ConfigEntry
+) -> list[DeviceEntry]:
+    dr = dr_get(hass)
+    return [
+        d
+        for d in dr.devices.values()
+        if entry.entry_id in d.config_entries
+        and any(dom == DOMAIN for dom, _ in d.identifiers)
+    ]
+
+
+# ---- generic start/stop switch ----
+class _KiaServiceSwitch(CoordinatorEntity, SwitchEntity):
+    _attr_has_entity_name = True
+    _attr_should_poll = False
+    _attr_assumed_state = True
+
+    def __init__(
+        self,
+        coordinator,
+        device: DeviceEntry,
+        name: str,
+        start_service: str,
+        stop_service: str,
+        start_data: dict[str, Any] | None = None,
+    ):
+        super().__init__(coordinator)
+        self._device = device
+        self._vin = _vin_from_device(device)
+        self._attr_device_info = DeviceInfo(identifiers={(DOMAIN, self._vin)})
+        self._attr_name = name
+        self._start = start_service
+        self._stop = stop_service
+        self._start_data = start_data or {}
+        self._is_on = False
+
+    @property
+    def unique_id(self) -> str:
+        safe = self._attr_name.lower().replace(" ", "_")
+        return f"{self._vin}_{safe}"
+
+    @property
+    def is_on(self) -> bool:
+        return self._is_on
+
+    async def _call(self, svc: str, data: dict[str, Any] | None = None):
+        await self.hass.services.async_call(
+            DOMAIN,
+            svc,
+            data or {},
+            blocking=True,
+            target={"device_id": [self._device.id]},
+        )
+        await self.coordinator.async_request_refresh()
+
+    async def async_turn_on(self, **kwargs) -> None:
+        await self._call(self._start, self._start_data)
+        self._is_on = True
+        self.async_write_ha_state()
+
+    async def async_turn_off(self, **kwargs) -> None:
+        await self._call(self._stop)
+        self._is_on = False
+        self.async_write_ha_state()
+
+
+# ---- concrete switches ----
+class KiaClimateSwitch(_KiaServiceSwitch):
+    def __init__(self, coordinator, device: DeviceEntry):
+        super().__init__(
+            coordinator,
+            device,
+            "Climate",
+            "start_climate",
+            "stop_climate",
+            {"duration": 10, "temperature": 21, "climate": True, "defrost": True},
+        )
+        self._attr_icon = "mdi:car-seat-heater"
+
+
+class KiaChargingSwitch(_KiaServiceSwitch):
```

Comment:
Ideally a description is used not a new class per switch.

Reply:
Code is updated, please take another look.